docs/GDPs/gdp-duck-storage.rst

```diff
@@ -0,0 +1,456 @@
+:code:`data`
+   returns :code:`data` attribute of the underlying ndarray
+:code:`alignment`
+   the value given in the constructor
+:code:`axes`
+   string of unmasked axes, e.g. :code:`"IJ"` for a 2d field spanning longitude and latitude but not the vertical.
+:code:`mask`
+   similar to axes, but a tuple of booleans. :code:`(True, True, False)` would be a 2d field spanning longitude and
+latitude but not the vertical
+
+:code:`default_origin`
+   the value given in the constructor indicating the grid point to which the memory is aligned.
+
+:code:`gpu`
+   boolean, indicating whether the storage has a gpu buffer
+
+ToDo
+
+Methods
+=======
+
+:code:`__array__`, :code:`__array_interface__` and :code:`__cuda_array_interface__`
+   where the former two are only supported for storages with an actual CPU buffer, the latter only for GPU-enabled
+   storages
+
+:code:`__deepcopy__` and :code:`copy` methods
+   allocate new buffers and copy the contents
+
+:code:`__getitem__`
+   dimensions, for which a certain index is selected are returned as masked, while slices do not reduce dimensionality.
+
+:code:`__setitem__`
+   :ref:`broadcasting: and device selection is equivalent to that of a unary ufunc with a provided output buffer.
+   For example, :code:`stor_out[:,3:5, 0] = stor2d` would be equivalent to
+   :code:`np.positive(stor2d, out=stor_out[:,3:5, 0]`)
+
+:code:`copy`
+   allocate new buffers and copy the contents
+
+The following methods are used to ensure one-sided modifications to CPU or GPU buffers of the
+`SoftwareManagedGPUStorage` are tracked properly. They are no-ops for all other storage classes, but are there so that
+user code can be backend-agnostic in these cases.
+
+The use of these methods should only be necessary, if a reference to the storage buffers is kept and modified outside
+of GT4Py.
+
+:code:`set_device_modified`, :code:`set_host_modified`, :code:`set_device_synchronized`
+   mark a buffer as modified, so that it can be synchronized before the respective other buffer is accessed.
+
+:code:`host_to_device` (:code:`device_to_host`)
+   Triggers a copy from host (device) buffer to the sibling in device (host) memory, if the host (device) is marked as
+   modified or the method is called with `force=True`. After a call to either of these methods, the buffers are flagged
+   as synchronized.
+
+:code:`synchronize`
+   Triggers a copy between host and device buffers if the host or device, respectively are marked as modified. The
+   buffers are marked as in sync as a consequence.
+
+
+Universal Functions
+^^^^^^^^^^^^^^^^^^^
+
+Universal functions, such as mathematical binary operations and logical operators are supported through the
+:code:`numpy.lib.mixins.NDArrayOperatorsMixin` type and the `__array_ufunc__` interface. We support the methods
+`__call__` and `reduce` of the numpy ufunc mechanism.
+
+If the :code:`reduce` method of ufuncs is used, this results in a Storage with the dimensions masked along which the
```

Comment:
This section is a bit obscure to me. I do not really know what this line is about..

Reply:
Ok, I will expand it. Will also provide some background on ufunc.reduce.

Reply:
I expanded the explanation a bit.